test/runtime/test_apply_requirement_modifications.rb

```diff
@@ -304,6 +304,202 @@ module Runtime
+                refute plan.syskit_has_async_resolution?
+                assert plan.syskit_pending_forced_resolution?
+
+                # Triggers new resolution because of the last 'force'
+                execute { Runtime.apply_requirement_modifications(plan) }
+                assert plan.syskit_has_async_resolution?
+                refute plan.syskit_current_resolution.cancelled?
+                refute plan.syskit_pending_forced_resolution?
+
+                # Resolution was not cancelled so we can join and apply it to the plan
+                plan.syskit_join_current_resolution
+                execute { Runtime.apply_requirement_modifications(plan) }
+                refute plan.syskit_has_async_resolution?
+
+                # From now on, all is well, this should not do anything
+                execute { Runtime.apply_requirement_modifications(plan) }
+                refute plan.syskit_has_async_resolution?
+            end
+
+            it "triggers a new resolution once with force: false when a force was " \
+               "used while a resolution was pending followed by another force: false" do
+                task_m = Composition.new_submodel
+                # `apply_requirement_modifications` never does anything if there are
+                # no requirements
+                requirement_task =
+                    plan.add_permanent_task(task_m.to_instance_requirements.as_plan)
+                requirement_task = requirement_task.planning_task
+                execute { requirement_task.start! }
+
+                # Finish one resolution, we're checking behaviour related to the force
+                # flag
+                execute { Runtime.apply_requirement_modifications(plan) }
+                begin
+                    plan.syskit_current_resolution.join
+                rescue Concurrent::CancelledOperationError # rubocop:disable Lint/SuppressedException
+                end
+                expect_execution { Runtime.apply_requirement_modifications(plan) }
+                    .to_emit(requirement_task.resolution_success_event)
+
+                execute { Runtime.apply_requirement_modifications(plan, force: true) }
+                assert plan.syskit_has_async_resolution?
+                refute plan.syskit_pending_forced_resolution?
+
+                FlexMock.use(plan) do |mock|
+                    mock.should_receive(syskit_finished_async_resolution?: false)
+                    # Cancels resolution, but does nothing else as
+                    # there is a pending unfinished resolution
+                    #
+                    # The test is to check that the `force` flag is remembered and will
+                    # be used the next time
+                    execute { Runtime.apply_requirement_modifications(plan, force: true) }
+                    execute do
+                        Runtime.apply_requirement_modifications(plan, force: false)
+                    end
```

Comment:
? I would try to have as little calls to apply_* as possible. If there are other conditions you want to check, make more tests for them,